### PR TITLE
feat(core): Upgrade API kubernetes 1.28 and controller-runtime to 0.16

### PR DESCRIPTION
**Pull request by gansheer**
Ref #5211 
Ref #5307 

Upgrade API kubernetes 1.28 and controller-runtime to 0.16

**Release Note**
```release-note
feat(core): Upgrade API kubernetes 1.28 and controller-runtime to 0.16
```

### CODE REVIEW

File: pkg/cmd/operator/operator.go [hunk 1]
```diff
@@ -201,22 +202,26 @@ func Run(healthPort, monitoringPort int32, leaderElection bool, leaderElectionID
 		}
 	}
 
+	defaultNamespaces := map[string]cache.Config{
+		operatorNamespace: {},
+	}
+
 	options := cache.Options{
-		ByObject: selectors,
```
Comment (squakez) [inline pkg/cmd/operator/operator.go:205]: We need to include those selector somewhere for caching reason. It seems this is no longer applied anywhere, or am I missing something?

Comment (gansheer) [reply pkg/cmd/operator/operator.go:205]: You are right, something got lost here. I will add it again.

**Comment by lburgazzoli** [comment]
> Is this expected to be merged for Camel K 2.3 ? It would be important to consider [openshift 4.14 uses kubernetes api 1.27](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html-single/release_notes/index#ocp-4-14-about-this-release).

According to the [kubernetes-version-compatibility](https://github.com/kubernetes-sigs/kubebuilder-release-tools/blob/master/VERSIONING.md#kubernetes-version-compatibility), it should still be compatible with Kubernetes API version N-3 or N-4

**Comment by gansheer** [comment]
> Is this expected to be merged for Camel K 2.3 ? It would be important to consider [openshift 4.14 uses kubernetes api 1.27](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html-single/release_notes/index#ocp-4-14-about-this-release).

For now this is for main, so post 2.3. Before backporting I need to run the test on Openshift 4.14.
 
> According to the [kubernetes-version-compatibility](https://github.com/kubernetes-sigs/kubebuilder-release-tools/blob/master/VERSIONING.md#kubernetes-version-compatibility), it should still be compatible with Kubernetes API version N-3 or N-4

The test needs to run on Openshift 4.14 because other dependency upgrade that come with it, like the one on client-go, could have an impact somewhere, and the compatibility is not 100% (see https://github.com/kubernetes/client-go#compatibility-client-go---kubernetes-clusters).

After this one is fully done there should be the upgrade for API 1.29 and controller-runtime 0.17.

**Comment by squakez** [comment]
I don't think we should backport this to 2.3 as it may pose some compatibility problems with the existing running platform. Ie, the user is running a 2.3.0 version and the upgrade to 2.3.1 should work on the very same platform he's running 2.3.0.

File: pkg/cmd/operator/operator.go [hunk 2]
```diff
@@ -201,22 +202,27 @@ func Run(healthPort, monitoringPort int32, leaderElection bool, leaderElectionID
 		}
 	}
 
+	defaultNamespaces := map[string]cache.Config{
+		operatorNamespace: {},
```
Comment (squakez) [inline pkg/cmd/operator/operator.go:206]: You're probably missing to include the `watchNamespace`

**Comment by squakez** [comment]
@gansheer may I suggest a different strategy? Whenever we upgrade dependency, we better do it one by one instead of doing one shot, in order to understand which is the dependency that really fails. IMO, the first thing to do is to upgrade kubernetes only. We can bump to 1.28 and later to 1.29. Once we are sure that all is stable, we can bump any other dependency.

**Pull request closed without merging**